lib/r10k/module_loader/puppetfile.rb

```diff
@@ -114,6 +115,13 @@ def add_module_metadata(name, info)
       # @param [String] forge
       def set_forge(forge)
         @forge = forge
+        if @allow_puppetfile_forge
+          logger.debug _("Using Forge from Puppetfile: %{forge}") % { forge: forge }
+          # This method will append a trailing slash to the string, in place, so we
+          # dup it to avoid different behavior in the value of our instance var
+          # when `allow_puppetfile_forge` is true vs false
+          PuppetForge.host = forge.dup
```

Comment:
The dup is probably unnecessary, seeing as we don't really use this instance var for anything that I can see. Lemme know what you think. Also, if we're not using it... do we need it at all?

Reply:
yeah, this is confusing to me.. I would be in favor of removing `@forge` and the various `set_forge` methods if they're not actually doing anything, since it sorta seems like they're not? but it's unclear. Or perhaps replace the `set_forge` methods with ones that actually do something.. I would have imagined something calling `set_forge(forge_from_puppetfile)` and then having `PuppetForge.host = @forge` somewhere where it's actually being used? But I don't really know, everything is a mess haha.

Reply:
As the comment at the top of this block says, `set_forge` is part of the Puppetfile parsing DSL. It is [called|https://github.com/puppetlabs/r10k/blob/4a832b7e0090ca2ec04c60f48ce43b98a6b16d5f/lib/r10k/module_loader/puppetfile/dsl.rb#L29] when the Puppetfile is [evaluated|https://github.com/puppetlabs/r10k/blob/4a832b7e0090ca2ec04c60f48ce43b98a6b16d5f/lib/r10k/module_loader/puppetfile.rb#L63-L64] by that DSL. Until now, it has set this instance variable, but been a noop since, so far as I could tell, the var isn't used (we were deliberately ignoring `forge` declarations in the Puppetfile).

The actual way to make the `forge` declaration be used is to set the `host` field for the `PuppetForge` gem. That's what we do when [reading it from settings|https://github.com/puppetlabs/r10k/blob/4a832b7e0090ca2ec04c60f48ce43b98a6b16d5f/lib/r10k/initializers.rb#L67]. I didn't delete the instance variable, in case something else is expecting it be set, but I think maybe that's not the case. This seems even more likely given there's no getter for it. I can try that and see what happens.

Reply:
yeah, I know the DSL has a method that calls this, but I didn't see anywhere calling the DSL's `forge` method, so wondering if that could be removed too.

Reply:
The DSL represents an implementation of how to evalute a Puppetfile as a Ruby script. So if a Puppetfile has
```
forge "my.awesome.forge

mod "puppetlabs/apache"
```
`forge` and `mod` there are Ruby function calls, where the definition is what's in the DSL class. So when `instance_eval` happens, we evaluate the Ruby script that is the Puppetfile in the context of the DSL class, which has those methods defined, effectively calling `forge`, then `mod`.

I was trying to look for a good example of using `instance_eval` this way, but it seems like most people use it to add methods to individual objects, instead of actually executing code directly like we're doing here. The docs might help though: https://www.rubydoc.info/stdlib/core/BasicObject:instance_eval

Reply:
oh right, ugh. didn't realize just `grep`ing wouldn't be enough 😅  but that makes sense.